packages/contracts-bedrock/src/dispute/AnchorStateRegistry.sol

```diff
@@ -23,105 +22,222 @@ import { ISuperchainConfig } from "interfaces/L1/ISuperchainConfig.sol";
+        if (_gameL2BlockNumber <= _anchorL2BlockNumber) {
+            revert AnchorGameIsNewer(_anchorL2BlockNumber, _gameL2BlockNumber);
+        }
+        _updateAnchorStateWithValidNewerGame(_game);
+    }
 
-        // Must be a valid game.
-        if (address(factoryRegisteredGame) != address(game)) revert UnregisteredGame();
+    function _updateAnchorStateWithValidNewerGame(IFaultDisputeGame _game) internal {
+        _anchorGame = _game;
+        emit AnchorGameSet(_game);
+    }
 
-        // No need to update anything if the anchor state is already newer.
-        if (game.l2BlockNumber() <= anchors[gameType].l2BlockNumber) {
+    function _maybeRegisterAnchorGameCandidate() internal {
+        IFaultDisputeGame _game = IFaultDisputeGame(msg.sender);
+        // game must not be invalid
+        if (!isGameMaybeValid(_game)) return;
+        // if the game is older than the anchor game, we don't need it
+        if (_game.l2BlockNumber() < _anchorGame.l2BlockNumber()) {
             return;
         }
+        anchorGameCandidates.push(_game);
+    }
 
-        // Must be a game that resolved in favor of the state.
-        if (game.status() != GameStatus.DEFENDER_WINS) {
-            return;
+    /// @notice Callable by FaultDisputeGame contracts to update the anchor state.
+    function tryUpdateAnchorState() external {
+        _maybeRegisterAnchorGameCandidate();
+        uint256 _anchorGameBlockNumber = _anchorGame.l2BlockNumber();
+        uint256 _gasStart = gasleft();
+        // TODO: add padding to ensure we don't run out of gas
+        while (_gasStart - gasleft() < tryUpdateAnchorStateGas) {
+            // if there are no candidates to consider, break
+            if (_anchorGameCandidateIndex == anchorGameCandidates.length) break;
+            // If the game's l2BlockNumber is older than that of anchorGame, seek ahead.
+            IFaultDisputeGame _anchorCandidate = anchorGameCandidates[_anchorGameCandidateIndex];
+            if (_anchorCandidate.l2BlockNumber() <= _anchorGameBlockNumber) {
+                // We can confidently seek past games that don't increase the anchor game l2BlockNumber.
+                _anchorGameCandidateIndex++;
+            } else {
+                if (_isGameFinalized(_anchorCandidate)) {
+                    // If the game is finalized but invalid, we should move past it
+                    if (!isGameValid(_anchorCandidate)) {
+                        _anchorGameCandidateIndex++;
+                    } else {
+                        // If the game is finalized and valid, let's use it.
+                        _updateAnchorStateWithValidNewerGame(_anchorCandidate);
+                        _anchorGameCandidateIndex++;
+                        _anchorGameBlockNumber = _anchorCandidate.l2BlockNumber();
+                    }
+                } else {
+                    // If the game is not finalized, we could consider checking some games ahead of it, but for draft
```

Comment:
keen to determine whether we need to consider an algo that seeks past unfinalized games a bit (without moving cursor)